core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedStreamSourceChannel.java

```diff
@@ -342,43 +342,43 @@ protected boolean isLastFrame() {
 
     @Override
     public void awaitReadable() throws IOException {
-        if(Thread.currentThread() == getIoThread()) {
+        if (Thread.currentThread() == getIoThread()) {
             throw UndertowMessages.MESSAGES.awaitCalledFromIoThread();
         }
-        if (data == null && pendingFrameData.isEmpty()) {
-            synchronized (lock) {
-                if (data == null && pendingFrameData.isEmpty()) {
-                    try {
-                        waiters++;
-                        lock.wait();
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new InterruptedIOException();
-                    } finally {
-                        waiters--;
-                    }
+        synchronized (lock) {
+            while (data == null && pendingFrameData.isEmpty()) {
```

Comment:
I'm not sure this is strictly necessary given awaitReadable and awaitWritable are allowed to return spuriously.

This change will also conflict with #842

Reply:
You are right @carterkozak - these methods can return spuriously. 
I had to check API definition first. I can close this PR evidently :)